mbuild/lib/recipes/polymer.py

```diff
@@ -182,25 +194,60 @@ def build(self, n, sequence="A", add_hydrogens=True):
 
         head_tail = [head_port, tail_port]
 
-        for i, compound in enumerate(self._end_groups):
-            if compound is not None:
-                if self._headtail[i] is not None:
-                    head_tail[i].update_separation(self._headtail[i])
-                self.add(compound)
-                force_overlap(compound, compound.labels["up"], head_tail[i])
-            else:
-                if add_hydrogens:
-                    hydrogen = H()
-                    # Defaut to 1/2 H-C bond len
-                    head_tail[i].update_separation(0.0547)
-                    hydrogen["up"].update_separation(0.0547)
-                    self.add(hydrogen)
-                    force_overlap(hydrogen, hydrogen["up"], head_tail[i])
-                else:
-                    # if None, hoist port to polymer level
-                    self.add(
-                        head_tail[i], self._port_labels[i], containment=False
+        if make_periodic:
+            if np.any([x is not None for x in self._end_groups]):
+                raise ValueError(
+                    "Keyword 'make_periodic' cannot be defined if 'end_groups' are provided."
+                )
+            z_axis_transform(
+                compound=self,
+                point_on_z_axis=head_port.pos,
+                point_on_zx_plane=tail_port.pos,
+            )
+            bond_length = 2 * np.max(
+                [
+                    np.sqrt(np.sum(np.square(atm1.pos - atm2.pos)))
+                    for (atm1, atm2) in self.root.bond_graph.edges(
+                        self["monomer[0]"]
                     )
+                ]
+            )
+            self.periodic = [
+                x > bond_length for x in head_port.pos - tail_port.pos
```

Comment:
Instead of comparing the distance of the ports to a single monomer bond_length to determine periodicity, we should be using the box length instead to see if it's a shorter distance within the box as opposed to across the box edges. Something like:
```
indexDict = {"x":0, "y":1, "z":2}
index = indexDict["axis"]
distance_along_axis_within_box = np.abs(head_port.pos[index] - tail_port.pos[index])
distance_along_axis_periodic = -1 * distance_along_axis_within_box + self.box[index]
if distance_along_axis_within_box > distance_along_axis_periodic:
    self.periodicity[index] = True # set periodicity to ensure bond stays periodic
else:
    raise Error("Polymer box is too large such that the perioic bond in the {axis} dimension does not result in a bond that stretches across the boundary. Please reset the box boundary to...")
self.add_bond(tuple(head_port.anchor, tail_port.anchor) # just add the bond without moving anything
self.remove(tail_port); self.remove(head_port) # remove extra ports
```

Reply:
At this point there is no box.

I suppose with that in mind this check seems pointless in general. Forcing the bond will add the connection regardless, then one could force minimize it to make a ring or whatever else they want to do.

Reply:
Using ``self.remove(tail_port)`` as you suggested results in:
```python
File [~/bin/mbuild/mbuild/compound.py:1043] Compound.remove(self, objs_to_remove)
   [1041]         ports_removed.add(obj)
   [1042]         self._remove(obj)
-> [1043] obj.parent.children.remove(obj)
   [1044] self._remove_references(obj)
   [1046] objs_to_remove = objs_to_remove - ports_removed

AttributeError: 'NoneType' object has no attribute 'children'
```

Reply:
Using ``self.add_bond(tuple(head_port.anchor, tail_port.anchor)`` as you suggested resulted in the following during visualization:
```python
KeyError                                  Traceback (most recent call last)
File [~/bin/mbuild/mbuild/compound.py:3649], in Compound._clone_bonds(self, clone_of)
   [3646] try:
   [3647]     # bond order is added to the data dictionary as 'bo'
   [3648]     newone.add_bond(
-> [3649]         (clone_of[c1], clone_of[c2]), bond_order=data["bond_order"]
   [3650]     )
   [3651] except KeyError:

KeyError: <C pos=([-0.0759  0.0018  0.0021]), 3 bonds, id: 140190920896368>
```

Reply:
> At this point there is no box.
> 
> I suppose with that in mind this check seems pointless in general. Forcing the bond will add the connection regardless, then one could force minimize it to make a ring or whatever else they want to do.

Well, with a compound you can always set a box. At any stage. So if you wanted to verify that your bond is truly periodic, you could set your box before doing the periodic wrapping, and that would be a good time to raise the error. If there isn't a box yet, you can't raise the error since we don't know enough to determine if that bond goes across the boundary.

Reply:
> Using `self.remove(tail_port)` as you suggested results in:
I was able to make it work fine locally. I can post the code for you on another branch if that helps. I think a bug you could be running into for that issue may be the remove_overlaps in combination with the self.remove. Since the remove_overlaps also removes the ports, you're essentially doing it twice. It needs to be:
```python
self.add_bond((head_port.anchor, tail_port.anchor)) # just add the bond without moving anything
self.remove(tail_port); self.remove(head_port) # remove extra ports
```

Reply:
> Using `self.add_bond(tuple(head_port.anchor, tail_port.anchor)` as you suggested resulted in the following during visualization:

Yep, you're right. The tuple was putting the particles into a tuple of a tuple. If you just replace the outside tuple call, it works fine.

Reply:
> At this point there is no box.
> 
> I suppose with that in mind this check seems pointless in general. Forcing the bond will add the connection regardless, then one could force minimize it to make a ring or whatever else they want to do.

I agree that we don't need this check and an error. For example, what I wanted this periodic bonding feature for was to simulating a lattice of polymers where the heads and tails are bonded together across the periodic boundary. Obviously the initial bond wasn't perfect, but a short energy minimization after-the-fact fixed it.